validation/validate.py

```diff
@@ -0,0 +1,268 @@
+                        required=True,
+                        help='Path to the image directory.')
+    parser.add_argument('--pairs_file_name',
+                        type=str,
+                        required=True,
+                        help='Filename of pairs.txt')
+    parser.add_argument('--face_verification_container',
+                        type=str,
+                        required=True,
+                        help='Name of docker container for face verification')
+    parser.add_argument('--num_folds',
+                        type=int,
+                        required=True,
+                        help='Number of cross validation folds')
+    distance_metrics = [f'{attr}'
+                        for attr in dir(DistanceMetric)
+                        if not callable(getattr(DistanceMetric, attr))
+                        and not attr.startswith("__")]
+    parser.add_argument(
+        '--distance_metric',
+        type=str,
+        required=True,
+        help=f"Distance metric for face verification: {distance_metrics}.")
+    parser.add_argument('--threshold_start',
+                        type=float,
+                        required=True,
+                        help='Start value for distance threshold.')
+    parser.add_argument('--threshold_end',
+                        type=float,
+                        required=True,
+                        help='End value for distance threshold')
+    parser.add_argument(
+        '--embedding_size',
+        type=int,
+        required=True,
+        help='Size of face vectors from face_verification_container.')
+    parser.add_argument(
+        '--threshold_step',
+        type=float,
+        required=True,
+        help='Step size for iterating in cross validation search.')
+    parser.add_argument(
+        '--subtract_mean',
+        action='store_true',
+        help=f"Subtract mean of embeddings before distance calculation.")
+    parser.add_argument(
+        '--divide_stddev',
+        action='store_true',
+        help=f"Divide embeddings by stddev before distance calculation.")
+    return parser.parse_args()
+
+
+def _main(args: Namespace) -> None:
+    args = _parse_arguments()
+    pairs = _read_pairs(args.pairs_file_name)
+    pair_paths, labels = _get_paths_and_labels(args.image_dir, pairs)
+    flat_paths = [path for pair in pair_paths for path in pair]
+    client = docker.from_env()
+    volumes = {args.image_dir: {'bind': '/images', 'mode': 'ro'}}
+    img_mount = ' '.join([f'/images/{path}' for path in flat_paths])
+    stdout = client.containers.run(args.face_verification_container,
+                                   img_mount,
+                                   volumes=volumes,
+                                   auto_remove=True,
+                                   environment=["PREALIGNED=true"])
+    response = json.loads(stdout.decode('utf-8').strip())
+    assert(max([len(i) for i in response['faceVectors']]) <= 1)
```

Comment:
It would be nice to support matching images that have more than one person in the image so that we can use the representative test set. This would require either aligning images at runtime so PREALIGNED=false and then the return type would be List[List[embedding]] and you would be comparing two List[embedding] to see if any of the pairs of embeddings match. The other option would be to run the facenet align-mtcnn.py script on all your images which saves images with multiple faces in this format ([align-mtcnn](https://github.com/armanrahman22/facenet/blob/d0498a194c926d4180a9513a356fda7a2580713d/facenet_sandberg/align/align_dataset_mtcnn.py#L176)). And then use files of that format as input and parse out the basename for matching faces that came from the same photo. I think option one is easier to implement, but might be less efficient. Lemme know if this makes sense to do!

Reply:
The way the validation script works right now is raw images -> align -> generate_pairs -> validate. Since the generate_pairs takes a small subset of the images per fold, would it be possible for me to just choose images in generate_pairs that are not group photos? That would be a simple check, and would not require changing the validation script at all.